samples/custom_module/dynamic/module.cc

```diff
@@ -0,0 +1,230 @@
+// CustomModuleState.
+//
+// Assumed thread-safe (by construction here, as it's immutable), though if any
+// mutable state is stored here it will need to be synchronized by the
+// implementation.
+class CustomModule final : public vm::NativeModule<CustomModuleState> {
+ public:
+  using vm::NativeModule<CustomModuleState>::NativeModule;
+
+  ~CustomModule() override {
+    iree_custom_module_basic_unregister_types(instance());
+  }
+
+  // Creates per-context state when the module is added to a new context.
+  // May be called from any thread.
+  StatusOr<std::unique_ptr<CustomModuleState>> CreateState(
+      iree_allocator_t host_allocator) override {
+    auto state = std::make_unique<CustomModuleState>(host_allocator);
+    return state;
+  }
+};
+
+}  // namespace
+
+// Creates a native custom module that can be reused in multiple contexts.
+// The module itself may hold state that can be shared by all instantiated
+// copies but it will require the module to provide synchronization; usually
+// it's safer to just treat the module as immutable and keep state within the
+// instantiated module states instead.
+//
+// Note that while we are using C++ bindings internally we still expose the
+// module as a C instance. This hides the details of our implementation and
+// is required for working across the dynamic library boundary.
+extern "C" IREE_VM_DYNAMIC_MODULE_EXPORT iree_status_t create_custom_module(
+    iree_vm_dynamic_module_version_t max_version, iree_vm_instance_t* instance,
+    iree_host_size_t param_count, const iree_string_pair_t* params,
+    iree_allocator_t host_allocator, iree_vm_module_t** out_module) {
+  // Ensure the version matches; the version will change if the VM module
+  // interface changes and existing libraries are incompatible.
+  if (max_version != IREE_VM_DYNAMIC_MODULE_VERSION_LATEST) {
+    return iree_make_status(
+        IREE_STATUS_UNIMPLEMENTED,
+        "unsupported runtime version %u, module compiled with version %u",
+        max_version, IREE_VM_DYNAMIC_MODULE_VERSION_LATEST);
+  }
+
+  // Ensure HAL types are available. We need to do this as we're being
+  // dynamically loaded and can't automatically access the hosting process
+  // variables.
+  IREE_RETURN_IF_ERROR(iree_hal_module_resolve_all_types(instance));
+
+  // Register custom types used by the module against the instance.
+  // Note that this function must be safe to call multiple times as the module
+  // may be loaded multiple times.
+  IREE_RETURN_IF_ERROR(iree_custom_module_basic_register_types(instance));
+
+  // Create the custom module and return it to the runtime.
+  // NOTE: this isn't using the allocator here and that's bad.
+  auto module = std::make_unique<CustomModule>(
+      "custom", /*version=*/0, instance, host_allocator,
+      iree::span<const vm::NativeFunction<CustomModuleState>>(
+          kCustomModuleFunctions));
+  *out_module = module.release()->interface();
+  return iree_ok_status();
```

Comment:
Could this use the allocator? The NOTE is ominous (could reword to say why it's bad in general / why it's okay for this style of C++ application)